mkldnn: relu, sum and batch norm

sreekanth-yalachigere opened the pull request:
**Activator Relu (eltwise)**

```
core/providers/mkldnn/activation
    activations.cc
    activations.h
```
Modified CPU Relu class definition to enable MKLDNN Relu class to inherit for fallback option

`class Relu final : public OpKernel { 
`to
`class Relu : public OpKernel {`

**Sum Kernel**

```
core\providers\mkldnn\math\sum.cc
core\providers\mkldnn\math\sum.h
```
Modified base class `Sum_6 `in `core\providers\cpu\math\element_wise_ops.h` Removed final
for mkldnn to fallback to cpu computation.

**Batch Norm Kernel**

```
core\providers\mkldnn\nn\batch_norm.cc
core\providers\mkldnn\nn\batch_norm.h
```
Modified base class `onnxruntime::BatchNorm` in file `core\providers\cpu\nn\batch_norm.h`
Removed final and made epsilon_ as protected member.

sreekanth-yalachigere commented:
@snnn @@jywu-msft 

Relu uninitialized memory warning fixed: MKLDNN V0.17.2
Sum: unexpected result: Fixed

jywu-msft commented on onnxruntime/core/providers/mkldnn/math/sum.cc:
```diff
@@ -0,0 +1,231 @@
+  const int num_dimensions;
+
+  SumParams(const std::vector<mkldnn::memory::dims>& dims,
+    const mkldnn::memory::dims& dst_dims, const int numinputs, 
```
can remove const from numinputs and dimensions parameters.

jywu-msft commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+  mkldnn::memory::dims& dst_dims;
+  size_t num_dimensions;
+
+  ReluParams(mkldnn::memory::dims& src_dims, mkldnn::memory::dims& dst_dims, 
```
src_dims and dst_dims references can be made const

jywu-msft commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+namespace {
+// Struct which encapsulates parameters for MKLDNN Pool primitive.
+struct ReluParams {
+  mkldnn::memory::dims& src_dims;
```
src_dims and dst_dims references can be made const

jywu-msft commented on onnxruntime/core/providers/mkldnn/nn/batch_norm.cc:
```diff
@@ -0,0 +1,287 @@
+    const mkldnn::memory::dims& scale_dims_mkl,
+    const mkldnn::memory::dims& b_dims_mkl, const mkldnn::memory::dims& mean_dims_mkl,
+    const mkldnn::memory::dims& var_dims_mkl, const mkldnn::memory::dims& dst_dims_mkl,
+    const float eps, const int dimensions)
```
eps and dimensions params don't need to be const as they are not references and will be copied.

sreekanth-yalachigere commented:
Re review: done.

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+  size_t num_dimensions;
+
+  ReluParams(const mkldnn::memory::dims& src_dims, const mkldnn::memory::dims& dst_dims, 
+		size_t dimensions = 0)
```
is dimensions intentionally a default parameter?

jywu-msft replied:
actually for dimensions, can't we compute it from the src_dims? same comment for all these kernels...

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/batch_norm.h:
```diff
@@ -0,0 +1,18 @@
+template <typename T>
+class BatchNorm final : public onnxruntime::BatchNorm<T> {
+ public:
+  BatchNorm(const OpKernelInfo& info) : onnxruntime::BatchNorm<T>(info) {}
```
Mark single parameter constructors explicit.

pranavsharma commented on onnxruntime/core/providers/mkldnn/math/sum.h:
```diff
@@ -0,0 +1,22 @@
+template <typename T>
+class Sum final : public onnxruntime::Sum_6<T> {
+ public:
+   Sum(const OpKernelInfo& info) : onnxruntime::Sum_6<T>(info) {}
```
Mark single parameter constructors explicit.

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.h:
```diff
@@ -0,0 +1,20 @@
+template <typename T>
+class Relu : public onnxruntime::Relu<T> {
+ public:
+  Relu(const OpKernelInfo& info) : onnxruntime::Relu<T>(info) {}
```
Mark single parameter constructors explicit.

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+  const TensorShape& x_shape = X->Shape();
+  const auto& x_dims = x_shape.GetDims();
+
+  if (X->Shape().NumDimensions() > 5 ) {
```
can you leave a comment here saying why we're defaulting to the non-mkldnn implementation?

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+
+  try {
+    ReluParams pool_params(src_dims_mkl, dst_dims_mkl, x_shape.NumDimensions());
+    ReluPrimitive<T>* relulPrimitive = ReluPrimitivePool<T>::Get(pool_params);
```
fix naming convention and typo.

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  std::unique_ptr<mkldnn::memory::desc> 
+	GetDstMemoryDesc() const { return context_.dst_md; }
```
this is not used any where. get rid of dead functions.

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+	GetDstMemoryDesc() const { return context_.dst_md; }
+
+  std::unique_ptr<mkldnn::eltwise_forward::primitive_desc> 
+	GetPrimitiveDesc() const {
```
is this used anywhere? if not, remove it.

pranavsharma commented on onnxruntime/core/providers/mkldnn/math/sum.cc:
```diff
@@ -0,0 +1,231 @@
+
+  ORT_ENFORCE(num_inputs > 0, "MKLDNN Sum kernel: Must have at least one input");
+
+  if (num_inputs == 1) {
```
please make a note why you're defaulting to non-mkldnn implementation here.

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/batch_norm.cc:
```diff
@@ -0,0 +1,287 @@
+  const mkldnn::memory::dims& mean_dims;
+  const mkldnn::memory::dims& var_dims;
+  const mkldnn::memory::dims& dst_dims;
+  const float epsilon;
```
why are epsilon and num_dimensions passed as const?

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/batch_norm.cc:
```diff
@@ -0,0 +1,287 @@
+
+  ~BatchNormPrimitive() = default;
+
+  void Compute(const T* src_data, const T* scale_data, const T* b_data, 
```
please follow similar guidelines as Relu. data that is intended to be mutated (like output data) should not be passed as const to begin with (to avoid const_cast later on).

pranavsharma commented on onnxruntime/core/providers/mkldnn/nn/batch_norm.cc:
```diff
@@ -0,0 +1,287 @@
+    context_.dst_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(dst_data)));
+
+    T* scaleShift_buf = static_cast<T*>(context_.scale_shift_mem->get_data_handle());
```
fix scaleShift naming convention.

sreekanth-yalachigere commented:
@pranavsharma @jywu-msft 
thanks for your review. addressing your review comments.

sreekanth-yalachigere commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -0,0 +1,204 @@
+
+  ~ReluPrimitive() = default;
+
+  void Compute(const T* src_data, const T* dst_data) {
```
@pranavsharma I looked at compute class of other kernels (conv, pool and lrn). In all cases, dst_data is const and const_cast is used.

pranavsharma replied:
it's rather unfortunate that this was missed in our code reviews before. we should do the right thing now.

sreekanth-yalachigere replied:
Can I leave it as const now? 
I'll make these changes in my next PR.

pranavsharma replied:
We should change the incorrect usage in this code. I'll fix the previous violations.

sreekanth-yalachigere commented:
addressed all the issues. 
removed num_dimensions and using src_dims.size() instead.
removed const _cast and const from dst_data
explicit for single parameter constructors
typo and naming conversions

pranavsharma commented on onnxruntime/core/providers/mkldnn/activation/activations.cc:
```diff
@@ -178,12 +169,12 @@ Status Relu<T>::Compute(OpKernelContext* context) const {
+    ReluPrimitive<T>* relul_primitive = ReluPrimitivePool<T>::Get(pool_params);
 
-    relulPrimitive->Compute(src_data, dst_data);
+    relul_primitive->Compute(src_data, dst_data);
```
typo: relu_primitive.

sreekanth-yalachigere replied:
sorry. Missed it.

pranavsharma commented:
LGTM 👍

pranavsharma commented:
@jywu-msft i guess this can be merged, unless you've any comments.
